In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('amazon_baby_Product.csv')

df.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  
0  These flannel wipes are OK, but in my opinion ...       3  
1  it came early and was not disappointed. i love...       5  
2  Very soft and comfortable and warmer than it l...       5  
3  This is a product well worth the purchase.  I ...       5  
4  All of my kids have cried non-stop when I trie...       5

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183531 entries, 0 to 183530
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   name    183213 non-null  object
 1   review  182702 non-null  object
 2   rating  183531 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 4.2+ MB


In [4]:
df.describe()

rating
count  183531.000000
mean        4.120448
std         1.285017
min         1.000000
25%         4.000000
50%         5.000000
75%         5.000000
max         5.000000

In [5]:
#Checking the number of Reviews for each product
df['name'].value_counts()

Vulli Sophie the Giraffe Teether                                                                                                                                                                                                                                                                                                                       785
Simple Wishes Hands-Free Breastpump Bra, Pink, XS-L                                                                                                                                                                                                                                                                                                    562
Infant Optics DXR-5 2.4 GHz Digital Video Baby Monitor with Night Vision                                                                                                                                                                                                                                          

#### From above, Vulli Sophie the Giraffe Teether has the highest review

In [6]:
#Checking for Missing Values in the dataset
df.isnull().sum()

name      318
review    829
rating      0
dtype: int64

#### There are 829 missing values in the review column and 318 missing values in the name column

In [7]:
len(df) - len(df.dropna())

1147

#### Whick makes a total 1,147 missing values in the dataset

In [8]:
len(df) #Total length of the dataset

183531

#### There is a total value of 183531 so deleting the NA values will still leave us with a reasonable amount of data

In [9]:
#Droping rows missing values
df.dropna()

name  \
0                                Planetwise Flannel Wipes   
1                                   Planetwise Wipe Pouch   
2                     Annas Dream Full Quilt with 2 Shams   
3       Stop Pacifier Sucking without tears with Thumb...   
4       Stop Pacifier Sucking without tears with Thumb...   
...                                                   ...   
183526  Baby Teething Necklace for Mom Pretty Donut Sh...   
183527  Baby Teething Necklace for Mom Pretty Donut Sh...   
183528   Abstract 2 PK Baby / Toddler Training Cup (Pink)   
183529  Baby Food Freezer Tray - Bacteria Resistant, B...   
183530  Best 2 Pack Baby Car Shade for Kids - Window S...   

                                                   review  rating  
0       These flannel wipes are OK, but in my opinion ...       3  
1       it came early and was not disappointed. i love...       5  
2       Very soft and comfortable and warmer than it l...       5  
3       This is a product well worth the purchase.  I ...       5  
4       All of my kids have cried non-stop when I trie...       5  
...                                                   ...     ...  
183526  Such a great idea! very handy to have and look...       5  
183527  This product rocks!  It is a great blend of fu...       5  
183528  This item looks great and cool for my kids.......       5  
183529  I am extremely happy with this product. I have...       5  
183530  I love this product very mush . I have bought ...       5  

[182384 rows x 3 columns]

#### All the missing values have been deleted which makes the size of the data now 182384

In [10]:
#Checking the review column for any impropper value
for i in range(0, len(df)-1):
    if type(df.iloc[i]['review']) != str:
        df.iloc[i]['review'] = str(df.iloc[i]['review'])

C:\Users\HP\Anaconda2\envs\tensorflow\lib\site-packages\pandas\core\series.py:1056: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cacher_needs_updating = self._check_is_chained_assignment_possible()


## Classifying the rating column into Good or Bad sentiment
#### Rating of 3 to 5 are Good and 1 to 2 is bad
#### 1 signifies Good and 0 signifies Bad

In [11]:
#A function to create sentiment based on ratings
def sentiment(n):
    return 1 if n >= 3 else 0

#Creating a new sentiment column
df['Sentiment'] = df['rating'].apply(sentiment)
df.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  Sentiment  
0  These flannel wipes are OK, but in my opinion ...       3          1  
1  it came early and was not disappointed. i love...       5          1  
2  Very soft and comfortable and warmer than it l...       5          1  
3  This is a product well worth the purchase.  I ...       5          1  
4  All of my kids have cried non-stop when I trie...       5          1

#### There is a new column called Sentiment which contains the values 1(for good) and 0(for bad)

In [12]:
#Checking the products with bad reviews
df[df['Sentiment'] == 0]

name  \
21        Nature's Lullabies Second Year Sticker Calendar   
41             SoftPlay Giggle Jiggle Funbook, Happy Bear   
50                              SoftPlay Cloth Book, Love   
70      Hunnt&reg; Falling Flowers and Birds Kids Nurs...   
71      Wall Decor Removable Decal Sticker - Colorful ...   
...                                                   ...   
183414  Samsung Baby Care Washer, Stainless Platinum, ...   
183461                    Mud Pie Milestone Stickers, Boy   
183463       Best BIB for Baby - Soft Bib (Pink-Elephant)   
183472       Bouncy&reg; Inflatable Real Feel Hopping Cow   
183503  Maxboost iPhone 5S/5 Case - Protective Snap-on...   

                                                   review  rating  Sentiment  
21      I only purchased a second-year calendar for my...       2          0  
41      This bear is absolutely adorable and I would g...       2          0  
50      This book is boring. Nothing to stimulate my g...       1          0  
70      The reason:Small sizeHard to apply on the wall...       1          0  
71      Would not purchase again or recommend. The dec...       2          0  
...                                                   ...     ...        ...  
183414  My infant goes to a really crappy daycare, and...       1          0  
183461  Pretty please open and inspect these stickers ...       1          0  
183463  Great 5-Star Product but An Obvious knock-off ...       1          0  
183472  When I received the item my initial thought wa...       2          0  
183503  I got this case in the mail today, it came on ...       2          0  

[26493 rows x 4 columns]

#### There are 26493 products with bad reviews 

In [13]:
#Checking the products with good reviews
df[df['Sentiment'] == 1]

name  \
0                                Planetwise Flannel Wipes   
1                                   Planetwise Wipe Pouch   
2                     Annas Dream Full Quilt with 2 Shams   
3       Stop Pacifier Sucking without tears with Thumb...   
4       Stop Pacifier Sucking without tears with Thumb...   
...                                                   ...   
183526  Baby Teething Necklace for Mom Pretty Donut Sh...   
183527  Baby Teething Necklace for Mom Pretty Donut Sh...   
183528   Abstract 2 PK Baby / Toddler Training Cup (Pink)   
183529  Baby Food Freezer Tray - Bacteria Resistant, B...   
183530  Best 2 Pack Baby Car Shade for Kids - Window S...   

                                                   review  rating  Sentiment  
0       These flannel wipes are OK, but in my opinion ...       3          1  
1       it came early and was not disappointed. i love...       5          1  
2       Very soft and comfortable and warmer than it l...       5          1  
3       This is a product well worth the purchase.  I ...       5          1  
4       All of my kids have cried non-stop when I trie...       5          1  
...                                                   ...     ...        ...  
183526  Such a great idea! very handy to have and look...       5          1  
183527  This product rocks!  It is a great blend of fu...       5          1  
183528  This item looks great and cool for my kids.......       5          1  
183529  I am extremely happy with this product. I have...       5          1  
183530  I love this product very mush . I have bought ...       5          1  

[157038 rows x 4 columns]

#### There are 157038 products with good reviews

## Feature Engineering

In [14]:
df[df['name'] == 'Vulli Sophie the Giraffe Teether']

name  \
34313   Vulli Sophie the Giraffe Teether   
34314   Vulli Sophie the Giraffe Teether   
34315   Vulli Sophie the Giraffe Teether   
34316   Vulli Sophie the Giraffe Teether   
34317   Vulli Sophie the Giraffe Teether   
...                                  ...   
159649  Vulli Sophie the Giraffe Teether   
159650  Vulli Sophie the Giraffe Teether   
159651  Vulli Sophie the Giraffe Teether   
159652  Vulli Sophie the Giraffe Teether   
159653  Vulli Sophie the Giraffe Teether   

                                                   review  rating  Sentiment  
34313   He likes chewing on all the parts especially t...       5          1  
34314   My son loves this toy and fits great in the di...       5          1  
34315   There really should be a large warning on the ...       1          0  
34316   All the moms in my moms' group got Sophie for ...       5          1  
34317   I was a little skeptical on whether Sophie was...       5          1  
...                                                   ...     ...        ...  
159649  My baby loves her Sophie Chew Toy. She can che...       5          1  
159650  Sophie the Giraffe was a big hit at the baby s...       5          1  
159651  quick shipping and perfect product. I would pu...       5          1  
159652  My baby who is currently teething love his Sop...       5          1  
159653  I know several baits that are enjoying the sof...       5          1  

[785 rows x 4 columns]

In [15]:
#REMOVE PUNCTUATIONS FROM THE DATA
import string
mess = 'Sample message! Notice: it has punctuation'

#Check characters to see if ther are in punctuation
nopunc = [char for char in mess if char not in string.punctuation]

#join the characters again to form the string
nopunc = ''.join(nopunc)

nopunc.split()

['Sample', 'message', 'Notice', 'it', 'has', 'punctuation']

In [16]:
#REMOVE WORDS THAT CONTAINS NUMBERS AND SPECIAL CHARACTERS
text = 'this nu234$@#4 has numbers 563$#&*'
text = [word for word in text if not any(c.isdigit() for c in word)] #removing numbers
text = ''.join(text)

red = [value for value in text if value.isalnum() or value == ' '] #removing special characters
red = ''.join(red)
red.split()

['this', 'nu', 'has', 'numbers']

In [17]:
#REMOVE STOPWORDS
from nltk.corpus import stopwords
stopwords.words('english')[0:10] #Show some basic stopwords

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]

In [18]:
#Removing stopwords from the data
clean_mess = [word for word in nopunc.split() if word.lower() not in stopwords.words('english')]
clean_mess

['Sample', 'message', 'Notice', 'punctuation']

### Putting it in a function and applying it to the Dataframe

In [24]:
'''
def text_process(mess):
    """
    Takes in a string of text, then preforms the following:
    1. Remove all punctuation
    2. Remove Numbers and Special characters
    3. Remove all stepwords
    4. Return a list of the cleaned text
    """
    
    #Check if numbers are part of the characters
    text = [word for word in mess if not any(c.isdigit() for c in word)]
    text = ''.join(text)
    #Check for special characters
    special = [value for value in text if value.isalnum() or value == ' '] #removing special characters
    special = ''.join(special)
    #Check character to see if there are in punctuation
    #nopunc = [char for char in str(mess) if char not in string.punctuation]
    nopunc = [word.strip(string.punctuation) for word in special.split(" ")]
    nopunc = ' '.join(nopunc)

    #Remove all Stepwords
    message = [word for word in nopunc.split() if word.lower() not in stopwords.words('english')]
    message = ' '.join(message)
    return message
    '''

In [36]:
def clean_text(text):
    
    """
    Takes in a string of text, then preforms the following:
    1. Remove all punctuation
    2. Remove Numbers and Special characters
    3. Remove all stepwords
    4. Return a list of the cleaned text
    """

    # tokenize text and remove puncutation
    text = [word.strip(string.punctuation) for word in str(text).split(" ")]
    # remove words that contain numbers
    text = [word for word in text if not any(c.isdigit() for c in word)]
    #remove special characters
    special = [value for value in text if value.isalnum() or value == ' ']
    special = ''.join(special)
    # remove stop words
    stop = stopwords.words('english')
    text = [x for x in text if x not in stop]
    # join all
    text = " ".join(text)
    return(text)

In [20]:
df.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  Sentiment  
0  These flannel wipes are OK, but in my opinion ...       3          1  
1  it came early and was not disappointed. i love...       5          1  
2  Very soft and comfortable and warmer than it l...       5          1  
3  This is a product well worth the purchase.  I ...       5          1  
4  All of my kids have cried non-stop when I trie...       5          1

In [37]:
df['cleaned_review'] = df['review'].apply(clean_text)

In [38]:
df.head(20)

name  \
0                            Planetwise Flannel Wipes   
1                               Planetwise Wipe Pouch   
2                 Annas Dream Full Quilt with 2 Shams   
3   Stop Pacifier Sucking without tears with Thumb...   
4   Stop Pacifier Sucking without tears with Thumb...   
5   Stop Pacifier Sucking without tears with Thumb...   
6             A Tale of Baby's Days with Peter Rabbit   
7   Baby Tracker&reg; - Daily Childcare Journal, S...   
8   Baby Tracker&reg; - Daily Childcare Journal, S...   
9   Baby Tracker&reg; - Daily Childcare Journal, S...   
10  Baby Tracker&reg; - Daily Childcare Journal, S...   
11  Baby Tracker&reg; - Daily Childcare Journal, S...   
12  Baby Tracker&reg; - Daily Childcare Journal, S...   
13  Baby Tracker&reg; - Daily Childcare Journal, S...   
14     Nature's Lullabies First Year Sticker Calendar   
15     Nature's Lullabies First Year Sticker Calendar   
16     Nature's Lullabies First Year Sticker Calendar   
17    Nature's Lullabies Second Year Sticker Calendar   
18    Nature's Lullabies Second Year Sticker Calendar   
19    Nature's Lullabies Second Year Sticker Calendar   

                                               review  rating  Sentiment  \
0   These flannel wipes are OK, but in my opinion ...       3          1   
1   it came early and was not disappointed. i love...       5          1   
2   Very soft and comfortable and warmer than it l...       5          1   
3   This is a product well worth the purchase.  I ...       5          1   
4   All of my kids have cried non-stop when I trie...       5          1   
5   When the Binky Fairy came to our house, we did...       5          1   
6   Lovely book, it's bound tightly so you may not...       4          1   
7   Perfect for new parents. We were able to keep ...       5          1   
8   A friend of mine pinned this product on Pinter...       5          1   
9   This has been an easy way for my nanny to reco...       4          1   
10  I love this journal and our nanny uses it ever...       4          1   
11  This book is perfect!  I'm a first time new mo...       5          1   
12  I originally just gave the nanny a pad of pape...       4          1   
13  I thought keeping a simple handwritten journal...       3          1   
14  Space for monthly photos, info and a lot of us...       5          1   
15  I bought this calender for myself for my secon...       4          1   
16  I love this little calender, you can keep trac...       5          1   
17  This was the only calender I could find for th...       5          1   
18  I completed a calendar for my son's first year...       4          1   
19  We wanted to get something to keep track of ou...       5          1   

                                       cleaned_review  
0   These flannel wipes OK opinion worth keeping  ...  
1   came early disappointed love planet wise bags ...  
2   Very soft comfortable warmer looks...fit full ...  
3   This product well worth purchase  I found anyt...  
4   All kids cried non-stop I tried ween pacifier ...  
5   When Binky Fairy came house special gift book ...  
6   Lovely book bound tightly may able add alot ph...  
7   Perfect new parents We able keep track baby's ...  
8   A friend mine pinned product Pinterest I decid...  
9   This easy way nanny record key events happen b...  
10  I love journal nanny uses everyday track daugh...  
11  This book perfect  I'm first time new mom book...  
12  I originally gave nanny pad paper write inform...  
13  I thought keeping simple handwritten journal w...  
14  Space monthly photos info lot useful stickers ...  
15  I bought calender second son It colorful room ...  
16  I love little calender keep track baby's first...  
17  This calender I could find second year I one f...  
18  I completed calendar son's first year really e...  
19  We wanted get something keep track child's mil...

## Creating a training feature
#### Combining both name and review column together to create a new column

In [39]:
df['Product'] = df['name'] + ' '+ df['cleaned_review']
df.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  Sentiment  \
0  These flannel wipes are OK, but in my opinion ...       3          1   
1  it came early and was not disappointed. i love...       5          1   
2  Very soft and comfortable and warmer than it l...       5          1   
3  This is a product well worth the purchase.  I ...       5          1   
4  All of my kids have cried non-stop when I trie...       5          1   

                                      cleaned_review  \
0  These flannel wipes OK opinion worth keeping  ...   
1  came early disappointed love planet wise bags ...   
2  Very soft comfortable warmer looks...fit full ...   
3  This product well worth purchase  I found anyt...   
4  All kids cried non-stop I tried ween pacifier ...   

                                             Product  
0  Planetwise Flannel Wipes These flannel wipes O...  
1  Planetwise Wipe Pouch came early disappointed ...  
2  Annas Dream Full Quilt with 2 Shams Very soft ...  
3  Stop Pacifier Sucking without tears with Thumb...  
4  Stop Pacifier Sucking without tears with Thumb...

In [40]:
#Checking for missing values again
df.dropna(inplace=True)

In [41]:
df.shape #The shape of the dataframe

(182384, 6)

#### Saving the new dataframe into a new csv file

In [42]:
#Only run this cell if changes are made to the data
#df.to_csv('cleaned_data.csv', index=False)